librad/src/git/trailer.rs

```diff
@@ -0,0 +1,273 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error {
+    #[error("The given input is not a well-formed trailer")]
+    IllFormed,
+
+    #[error("Could not parse the trailer due to an invalid token found")]
+    Token(#[from] InvalidToken),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer<'a> {
+    token: Token<'a>,
+    value: &'a str,
+}
+
+impl<'a> Trailer<'a> {
+    #[allow(dead_code)]
+    /// Try and parse a [Trailer] form a given raw &str that is
+    /// expected to represent a trailer in raw form.
+    pub fn parse(s: &'a str, separators: &'a [char]) -> Result<Trailer<'a>, Error> {
+        let parts: Vec<&str> = s.split(&separators[..]).map(|s| s.trim()).collect();
+        match parts.as_slice() {
+            [token, value] => Self::new(*token, *value),
+            _ => Err(Error::IllFormed),
```

Comment:
This variant accounts for the whole line not being a well-formed trailer, ie, we couldn't parse a `<token><separator><value>` from it. So if parsing that format fails, we report that the input is ill-formed. Does that change your perspective?

Reply:
Ya, but there can be multiple lines in a trailer, right? How will I know which line failed? Imagine a compiler that told you that you had a parse error but not which line :smile:

Reply:
I am not sure if you are referring to the RFC822 (see https://github.com/radicle-dev/radicle-link/pull/271#pullrequestreview-464834027) or suggesting preserving the info about which trailer as ill-formed.

Say we have:

```
Subject

Description

TokenA: ValueA
TokenB ValueB
TokenC: ValueC
```

Is your suggestion that the error should inform that the trailer line number #2 is ill-formed, or, say we have:

```
Subject

Description

TokenA: 
  Value1
  Value2
  Value3
```

... that we should inform which of the lines is ill-formed within a specific trailer is ill-formed? 

If the first, that can be useful indeed. Alternatively, we can include the str line that failed to parse in the error itself. If the second case,  I think that only the first line can be ill-formed, but I need to digest the contents of RFC822 better.

Reply:
> suggesting preserving the info about which trailer is ill-formed.

I'm suggesting this. So if we're parsing the error would be `IllFormed` and it could be any of the 3 trailer lines and, exactly, you're saying the second line is ill-formed. Which I actually missed as I was scanning through your comment :joy: 

In the second case, we might want a nested error, like the "For '{Token}' the '{Value}' was ill-formed".

I actually haven't read the RFC so not sure if the second suggestion is useful :grimacing:

Reply:
Cool, sounds good to me :)